homework_solution.md

```diff
@@ -0,0 +1,474 @@
+
+   ### Steps
+
+   1. Create an index on `from_station_id` and `to_station_id` on `trips` table to help query on those
+
+      ```
+      CREATE INDEX idx_from_station ON trips USING btree (from_station_id);
+      ```
+
+      ```
+      CREATE INDEX idx_to_station ON trips USING btree (to_station_id);
+      ```
+
+   2. Delete the exist `stations` table and recreate it with serial ids, so we can rely on auto-increment. (We could try and alter the table, but that seems to be an involved process, and our table being currently empty, it's easier to just recreate it)
+
+      ```
+      DROP TABLE stations;
+      ```
+
+      ```
+      CREATE TABLE stations
+      (
+         id SERIAL,
+         name VARCHAR(100)
+      );
+      ```
+
+      The effect of this can be perceived by running `SELECT nextval(pg_get_serial_sequence('stations', 'id'));`.
+      On the old table, that sequence, though present, wouldn't return a value. Now it does.
+
+   3. Get the first name for stations, assuming that as the correct one where there's more than one, combining the `from` and `to` column results
+
+      ```
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id
+      UNION
+      SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+      FROM trips
+      WHERE from_station_id IS NOT NULL
+      GROUP BY from_station_id;
+      ```
+
+   4. Insert into the `stations` table the result of our query
+
+      ```
+      INSERT INTO stations (
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+         UNION
+         SELECT from_station_id, (ARRAY_AGG(DISTINCT from_station_name))[1]
+         FROM trips
+         WHERE from_station_id IS NOT NULL
+         GROUP BY from_station_id
+      );
+      ```
+
+   5. Adjust the `id` sequence to be the highest `id` from our `INSERT` above
+
+      ```
+      SELECT setval(pg_get_serial_sequence('stations', 'id'), (SELECT MAX(id) FROM stations));
+      ```
+
+   6. Finally, add entries for records where a name exists, but the id is null, relying on the seqquence to generate an id. This has to be done in 2 steps, so we can exclude the records from the first insertions when we run the second one, avoiding duplicates.
+
+      ```
+      INSERT INTO stations (
+         SELECT nextval(pg_get_serial_sequence('stations', 'id')), from_station_name
+         FROM trips
+         WHERE
+            from_station_id IS NULL
+         AND
+            from_station_name NOT IN (SELECT name from stations)
```

Comment:
[question]
This is going to compare every name it finds to every name in the stations table. How might you avoid that? (hint: It's often easier to ask forgiveness than permission)

Reply:
The only way that comes to mind right now would be a `JOIN`, checking for nulls on that join. 
However, joining on the `name` feels wrong, and we don't have the `id` yet to rely on, so I'm at a loss here. (But open to suggestions!!)

Reply:
I was thinking more along the lines of putting a uniqueness constraint on the `station.name` column, and adding `ON CONFLICT DO NOTHING` to the end of the query. I don't know if that would work in this case though, because it's the bulk insert.

This is an 'easier to ask forgiveness' approach, because you're trying to just insert everything, end relying on an exception (conflict) to let you know when that won't work.  For context, this is a common paradigm in python communities (where most of my experience comes from).

Reply:
I didn't know that was a thing! Thanks for the reference.

(By the way, I was asked someone about this after posting the last comment, and their recommendation was actually to do the whole `INSERT` in a single step, relying on `COALESCE` to handle cases where the `id` is null. This way, we wouldn't need to run a check on what's already present in the `stations` table as we're doing here. Seemed pretty smart!!